JAB17 Allow to check 3rd party extensions compatibility before J! upgrade

piotrmocko opened the pull request:
JAB17 make it happen!

### Summary of Changes

Added methods to check 3rd party extensions (except languages) compatibility before J! upgrade.
This code is based on: https://github.com/nicksavov/joomla-cms/tree/2.5.x-with-pre-upgrade-check

It is a task which we are implementing on JAB 2017 with @wilsonge 

After applying a similar code to the above brunch it gives a following result where the compatibility is not checked against the manifest data stored in Joomla but against the update server of the extension.

But this PR is only a small part which has to be migrated from above brunch to J!3. The remaining part of code should have been done by other JAB17 team members and committed in another PR.

So do not expect to get the result as on this screen shot. It is only a visualization of whole feature.

![screenshot-127 0 0 1-2017-06-04-08-56-08](https://cloud.githubusercontent.com/assets/977916/26759652/33de5b02-4904-11e7-9964-4474cc0e0f29.png)

### Testing Instructions

Here are testing instructions for this PR, but it would be easier to test it when the remaining PRs will be committed.

Install the extension on your Joomla https://extensions.joomla.org/extensions/extension/site-management/seo-a-metadata/link-with-article-images-on-facebook/
Call the method getExtensions from the class JoomlaupdateModelDefault with an argument:
* '3.7.2' - there is  an update server and an update, so it is compatible
* '3.8.0' - there is an update server but there is no update, so it is not compatible
* '3.7.2' - remove the update server from database, there is no update server, so the result is NA.

In the file administrator/components/com_joomlaupdate/views/default/view.html.php
after `$model = $this->getModel();`
add this code with one of above arguments: `var_dump($model->getExtensions('3.7.2')); die;`

Core extensions and (core and 3rd party) language extensions and language packs are excluded from the check for compatibility.

Some more test examples: https://github.com/joomla/joomla-cms/pull/16494#issuecomment-313920961

### Expected result

You will get an array with extensions grouped into categories and above extension will be in following category for each scenario:
* compatible
* not_compatible
* na

### Actual result

No results as this feature does not exist

### Documentation Changes Required

None

piotrmocko committed:
Allow to check 3rd party extensions compatibility before J! upgrade

Added methods to check 3rd party extensions compatibility before J! upgrade.

JAB17 make it happen!

piotrmocko commented:
This pull request requires also https://github.com/joomla/joomla-cms/pull/16493

piotrmocko commented:
After whole code will be migrated, then this approach should be changed, that each extension will be checked against J! upgrade compatibility with a single AJAX request, to make sure that we will not end with a timeout when checking multiple extensions.

Bakual commented:
If I understand this right, we had something similar in J2.5 to show if an extension will be compatible with J3: https://github.com/joomla/joomla-cms/pull/5015

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+	 *
+	 * @return  mixed  An array of data items.
+	 *
+	 * @since   3.8.0
```
please use `__DEPLOY_VERSION__`

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+	 *
+	 * @return  array  An array of data items.
+	 *
+	 * @since   3.8.0
```
please use `__DEPLOY_VERSION__`

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+	 * 
+	 * @return array
+	 *
+	 * @since 3.8.0
```
please use `__DEPLOY_VERSION__`

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+
+				// If there is a download URL then there is probably a compatible update
+				$download_url = $update->get('downloadurl');
+				if (!empty($download_url) && !empty($download_url->_data)) {
```
please move the `{` into its own line

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+				continue;
+			}
+
+			foreach ($locations as $location) {
```
please move the `{` into its own line

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+			$locations = $this->getUpdateSitesLocations($extension->extension_id);
+
+			// If there are no update servers then we do not know if there is a compatible update
+			if (empty($locations)) {
```
please move the `{` into its own line

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+
+		preg_match('/^(\d+\.\d+\.(\d+))/', $latest_version, $version);
+
+		foreach ($extensions as $extension) {
```
please move the `{` into its own line

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+	private function checkCompatibility($extensions, $latest_version)
+	{
+		// If empty, return the current value
+		if (empty($extensions) || !is_array($extensions)) {
```
please move the `{` into its own line

zero-24 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,114 @@ public function removePackageFiles()
+	 *
+	 * @param   string $latest_version The Joomla! version to test against
+	 *
+	 * @return  mixed  An array of data items.
```
the return type is a array? Or where i can be something different?

piotrmocko replied:
That is not my part of code but I will change that

zero-24 reviewed (dismissed):
Se also: http://213.160.72.75/joomla/joomla-cms/5370

zero-24 commented:
I'm not getting the point of that feature here. Is'nt that what the update server should take care of. So you can propose a different update definition of a different joomla version. So no need for extra coding?

piotrmocko commented:
It is possible that it is a similar code to PR #5015. The difference is that it is displayed in Joomla Update component and now updates are checked against the update server and I do not know how it was done in PR #5015.

This feature is a step to make a view in Joomla Update component to show the compatibility of 3rd party extensions against Joomla version which you are going to update to or upgrade.

This is a task which a got from @wilsonge on JAB17.

piotrmocko committed:
Code formatting

piotrmocko committed:
Code formatting

piotrmocko committed:
Code formatting

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,119 @@ public function removePackageFiles()
 	}
+
+	/**
+	 * Method to get a list with 3rd party extensions, sorted by compatibility
```
==>
of 3rd party

brianteeman commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,119 @@ public function removePackageFiles()
+	}
+
+	/**
+	 * Get an array with URL's to update servers for a given extension ID
```
==>
URLs

richard67 commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,119 @@ public function removePackageFiles()
+		$query->select('*')
+			->from($db->qn('#__extensions'))
+			->where($db->qn('protected') . ' = ' . $db->q('0'))
+			->where($db->qn('extension_id') . ' > ' . $db->q('9999'));
```
@wilsonge I have found some other PR which checks against 999 and not 9999: #16474  , see this comment: [https://github.com/joomla/joomla-cms/pull/16474#pullrequestreview-41929419](https://github.com/joomla/joomla-cms/pull/16474#pullrequestreview-41929419)
Can you check and let us know which is the right one, 999 or 9999?

wilsonge committed:
Merge branch 'staging' into patch-29

Bakual commented:
>  I do not know how it was done in PR #5015.

In the 2.5 extension manager, it has shown in a tooltip next to the version information if there is an update available for the same extension in J3. The information was stored in the "system_data" column of the #__extension table.

piotrmocko committed:
Fixes in comments

franz-wohlkoenig commented:
@piotrmocko can you give Example for each of the three Extensions neede for  Test?

Ruud68 commented:
I have a package containing of a component, a module and a plugin that work from 3.2 and up. In the package I also have a fields plugin that is installed as part of the package but only works on 3.7 and higher (it is not required for the correct working of the component)

My update server has the following: 3.[234567] That is for the whole package.

in the fields plugin I have the following entry:
<extension type="plugin" version="3.7" group="fields" method="upgrade">

What I understand is that this information (version="xyz" would be used (future) to do just what you propose: not install that plugin / component / module on joomla sites that have a lower version number.
Currently that is not implemented and no matter what you fill in to version in the plugin's manifest file, it will install on all versions of Joomla.

So just my two cents: if you want to be more in control of when to install what you probably need a combination of checking the update server AND the manifest version information.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

piotrmocko commented:
@franz-wohlkoenig I will give you tomorrow some extensions to test. 

@Ruud68 this feature is to give a Joomla administrator information if his extensions will be compatible with the version of Joomla which he want to install. 

@wilsonge did JAB17 team manage to migrate the code? I can help you with that if there is anything more to do.

piotrmocko commented:
@franz-wohlkoenig I have updated the description how to test this PR

franz-wohlkoenig commented:
@piotrmocko so there are 2 Test on 3.7.3-dev (with and -out Update-Server) and one on 3.8.0.

piotrmocko commented:
You can test it even on Joomla 3.5 if the code did not change to much. Those arguments which I have suggested are only telling the method to which Joomla you would like to update. I advise to test it on Joomla 3.7. But remember that there is also required PR #16493

franz-wohlkoenig commented:
For Tester: #16493 is merged, test on staging or latest nightly Build.

richard67 commented:
@wilsonge What is correct now to check for core extensions? Item ID 9999 like done in this PR here? Or 999 like done in PR #16474 ?
I ask you because this PR here seems to be based on your input.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

Bakual commented:
See https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L481
First extension ID installed regulary is 10000

Sidenote: that only works if the core extensions are installed using the Joomla Update component. If it's done using extension discovery after an unsupported manual upgrade, then core extensions have IDs higher than 10000.

richard67 commented:
Hmm, @Bakual ' s sidenode above makes me think it is not really save to use the ID number to check if an extension is core or not. But what else way do we have beside a hard-coded list which would have to be maintained every time an extension is added?

zero-24 commented:
> to check if an extension is core or not.

No it is not ;) I guess we need to hardcode a array (the best would be a global place) with the core extensions or implement a database flag `core` vs 3rd party.

richard67 commented:
@zero-24 Well, I agree. But that should be not scope of this PR, I think. Which method (hardcoded array or db column in the extensions table being set by core installation sql scripts) should be discussed in PLT, I think, because it is an important decision. Should I make a "discussion" issue for that (or RFC or however to call it) in the issue tracker?

richard67 commented:
@zero-24 Meanwhile I think the db column thing would be easier to maintain because no edit of list necessary, only have to set the right value of that column in the sql scripts for installation. Disadvantage: If some bad 3rd party extensions developer sets that value, too, in his/her sql script, it will cause trouble.

zero-24 commented:
@richard67 I think we can open a RFC (no code required yet) issue describing the problem and the solution and I can pass that to the maintainer chat on glip if needed.

But i think this here and the other one needs to be on hold until we have a clean solution.

zero-24 commented:
Can you do the RFC Issue? @richard67

richard67 commented:
Yes. Am on the way.

richard67 commented:
@zero-24 @Bakual I created the RFC issue #16579 to discuss the core or not core determination thing.

zero-24 commented:
Thanks

piotrmocko commented:
@zero-24 I have already applied changes according to your review

franz-wohlkoenig commented:
@piotrmocko applied PR and Plugin installed got:
![bildschirmfoto 2017-06-09 um 13 20 14](https://user-images.githubusercontent.com/8235763/26973374-8d6c77d8-4d16-11e7-8b00-a1b52dbac06b.jpg)
So i didn't get a Screen like in your first Comment.

piotrmocko commented:
My PR is only a part of what you see on the screen shot.

The remaining part of code should have been made by other volunteers on JAB17 
and @wilsonge was leading that team so he knows what is the progress.

franz-wohlkoenig commented:
@piotrmocko can you please update your Testing Instructions with this Informations so willing Testers know that ist not to Test now?

richard67 commented:
Regarding the question on how to detect if an extension is core or not which I discussed in my comments above, this PR here should not wait for a solution.

Let it continue to use the 9999.

Any later PR implementing a safe check will also have to look for such things and replace them by using the new way which it implements.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

richard67 committed:
Safe check for excluding core extensions

Use the new JExtensionHelper class to check if not core extension instead of the unsafe check for extension_id > 9999.

richard67 commented:
@piotrmocko I have created a PR against your branch which replaces the unsafe test for extension_id > 9999 to exclude core extensions by using the new JExtensionHelper class which has been merged in staging today:

[https://github.com/piotrmocko/joomla-cms/pull/1](https://github.com/piotrmocko/joomla-cms/pull/1 "https://github.com/piotrmocko/joomla-cms/pull/1")

I have not tested my change yet. I will do soon and let you know about the result.

Then you can accept it, which means you merge it into your branch patch-29.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

piotrmocko committed:
Merge pull request #1 from richard67/patch-2

Safe check for excluding core extensions

piotrmocko commented:
@richard67 I have merged your PR into mine

richard67 commented:
Good. I will test yours tonight when back from work. Do you have a code snippet we can use for test?

piotrmocko commented:
I have updated the test instruction with some code.

richard67 commented:
I have to test tomorrow or on weekend, today I am too tired now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

alikon commented on administrator/components/com_joomlaupdate/models/default.php:
```diff
@@ -995,4 +995,131 @@ public function removePackageFiles()
+
+		$query->select('*')
+			->from($db->qn('#__extensions'))
+			->where($db->qn('protected') . ' = ' . $db->q('0'));
```
the protected field is declared as an INTEGER not as a string
`protected  tinyint(3) NOT NULL DEFAULT 0,`

richard67 replied:
@piotrmocko You just change `' = ' . $db->q('0')` to `' = 0'` and it is ok.

piotrmocko committed:
Update default.php

richard67 commented:
@piotrmocko Sorry, just tested and found a silly mistake which I have made in my previous PR for you. I have forgotten to change variable name at some place. You will see that quickly when checking my new PR [https://github.com/piotrmocko/joomla-cms/pull/2](https://github.com/piotrmocko/joomla-cms/pull/2 "https://github.com/piotrmocko/joomla-cms/pull/2") for you.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

richard67 commented:
I tested meanwhile with the correction of my silly mistake (see my previous post). The test code produces now following warning:

getExtensions('3.7.2') Warning: Creating default object from empty value in /path_to_joomla/htdocs/test2/libraries/joomla/updater/update.php on line 267 Warning: Creating default object from empty value in /path_to_joomla/htdocs/test2/libraries/joomla/updater/update.php on line 267.

I think this comes now from your original code and not my corrections.

I will later try to find out more about that, but maybe you can check, too.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

richard67 commented:
The warning comes from calls "$update->loadFromXML($location);" with $location = "https://update.joomla.org/language/translationlist_3.xml".<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

richard67 commented:
Can it be that language packs in general need a special treatment for getting the update xml location from the translationlist xml?
The exclusion of core extensions from the list only excludes the core language pack (EN-GB) but not additional languages being installed on a multilingual site (which I used for testing).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

wojsmol commented:
@richard67 Lang packs update server is addef tu package but returns collecdtion type xml.

richard67 commented:
@wojsmol That's what I meant. So it seems @piotrmocko has to add extra handling for language packs to do the second step. As before I can try to help because I am interested in that feature, but I do not have much time.

richard67 commented:
Maybe the simple solution for the problem I found is to exclude language packs from the check, like it is already done with core extensions?

This could make sense because language packs are always coming up with new versions after the new Joomla version has been released, they do not prepare in advance versions compatible with future Jooma versions like other extensions do.

@piotrmocko Has @wilsonge said anything about that when giving you this task? Shall language packs be checked or not? For the reason I mentioned above they should not be checked in my opinion. Do you agree?

@piotrmocko If they shall not be checked, simply add "->where($db->qn('type') . ' <> ' . $db->q(language))" after the "->where($db->qn('protected') . ' = 0')".

But you should also merge my PR with the correction of my silly copy and paste mistake, otherwise nothing works and you get an SQL error.

richard67 commented:
The more I think about it, the more I think that language packs should be excluded from the check, like core extensions already are. I make another PR to @piotrmocko 's branch to help with that.

@piotrmocko Let me know if my PRs are not welcome. I do not want to takeover your PR, it is still your PR and you did most of the work. I only want to help with making it work because I am interested in the feature it helps to implement, and so I try to help as much as I can, but if your task was to do it without any help let me know and I will stop.

richard67 commented:
Hmm, I just see language packs are not so easy to exclude, because they have also a package (type = 'package') extension.

richard67 committed:
Exclude language packs and fix my copy paste error

1. Exclude language extensions and their parent packages in general because they never serve compatible updates for a new release in advance.
2. Correct my silly copy and paste error with forgotten rename of variable which I also did in PR [https://github.com/piotrmocko/joomla-cms/pull/2](https://github.com/piotrmocko/joomla-cms/pull/2)

richard67 commented:
@piotrmocko I've made a new PR to your branch, which excludes language extensions and their parent package from the check and includes also my previous correction. Please check [https://github.com/piotrmocko/joomla-cms/pull/3](https://github.com/piotrmocko/joomla-cms/pull/3 "https://github.com/piotrmocko/joomla-cms/pull/3"). I have already tested your PR with my corrections, and it seems to be ok.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

piotrmocko committed:
Merge pull request #3 from richard67/patch-5

Exclude language packs and fix my copy paste error

richard67 committed:
Add missing parantheses

Add missing parantheses around the OR condition so SQL does the right thing. Otherwise core extensions and language packs will not be properly excluded.

piotrmocko committed:
Merge pull request #4 from richard67/patch-6

Add missing parantheses

richard67 commented:
I have tested this item :white_check_mark: successfully on 2300d9770b92d1e186350f5d2a63b8725dcadf4e<br /><br />Details about test result and how I tested I will post in following comment.
@piotrmocko Please modify your testing insctructions and the description of your PR so it is clear that core extensions and (core and 3rd party) language extensions and language packs are excluded from the check for compatibility.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

piotrmocko commented:
@richard67 done

richard67 commented:
I inserted following code snippet between lines "$model = $this->getModel();" and "$this->loadHelper('select');" in file "administrator/components/com_joomlaupdate/views/default/view.html.php" for testing:
```
echo "getExtensions('3.7.4')<br /><br />";
$testExtensionsLists = $model->getExtensions('3.7.4');
echo "Compatible:<br />";
foreach ($testExtensionsLists['compatible'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br />  Not compatible:<br />";
foreach ($testExtensionsLists['not_compatible'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br />NA:<br />";
foreach ($testExtensionsLists['na'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br /><br />getExtensions('3.8.0')<br /><br />";
$testExtensionsLists = $model->getExtensions('3.8.0');
echo "Compatible:<br />";
foreach ($testExtensionsLists['compatible'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br />  Not compatible:<br />";
foreach ($testExtensionsLists['not_compatible'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br />NA:<br />";
foreach ($testExtensionsLists['na'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br /><br />getExtensions('4.0.0')<br /><br />";
$testExtensionsLists = $model->getExtensions('4.0.0');
echo "Compatible:<br />";
foreach ($testExtensionsLists['compatible'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br />  Not compatible:<br />";
foreach ($testExtensionsLists['not_compatible'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
echo "<br />NA:<br />";
foreach ($testExtensionsLists['na'] as $testExtension)
{
echo $testExtension->extension_id . ' ' . $testExtension->type . ' ' . $testExtension->name . '<br />';
}
```

The result was:
```
getExtensions('3.7.4')

Compatible:
10055 plugin plg_installer_webinstaller
10060 component COM_OSMAP
10062 plugin plg_system_ossystem
10109 library Alledia Framework (Don't uninstall this if you're using an Alledia/Joomlashack extension)
10110 component com_patchtester
10111 plugin PLG_PWEB_FBARTICLEIMAGES

Not compatible:

NA:
10044 template protostar-clone
10045 component com_phocaguestbook
10079 plugin PLG_OSMAP_JOOMLA
10086 file Komponente Phoca Guestbook - Deutsche Sprachdateien


getExtensions('3.8.0')

Compatible:
10055 plugin plg_installer_webinstaller
10060 component COM_OSMAP
10062 plugin plg_system_ossystem
10109 library Alledia Framework (Don't uninstall this if you're using an Alledia/Joomlashack extension)
10110 component com_patchtester

Not compatible:
10111 plugin PLG_PWEB_FBARTICLEIMAGES

NA:
10044 template protostar-clone
10045 component com_phocaguestbook
10079 plugin PLG_OSMAP_JOOMLA
10086 file Komponente Phoca Guestbook - Deutsche Sprachdateien


getExtensions('4.0.0')

Compatible:
10109 library Alledia Framework (Don't uninstall this if you're using an Alledia/Joomlashack extension)

Not compatible:
10055 plugin plg_installer_webinstaller
10060 component COM_OSMAP
10062 plugin plg_system_ossystem
10110 component com_patchtester
10111 plugin PLG_PWEB_FBARTICLEIMAGES

NA:
10044 template protostar-clone
10045 component com_phocaguestbook
10079 plugin PLG_OSMAP_JOOMLA
10086 file Komponente Phoca Guestbook - Deutsche Sprachdateien
```

richard67 commented:
Why the plugin PLG_OSMAP_JOOMLA is listed among the not available ("NA") is not clear to me yet. But it does not have a parent package so that cannot be the reason.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

richard67 commented:
So or so the result is correct for me. The PLG_OSMAP_JOOMLA in my test is updated with the component, and it has no parent package, so it is ok that it is "NA".

More testers please.

wojsmol commented:
@richard67 You assumption from https://github.com/joomla/joomla-cms/pull/16494#issuecomment-313911107 is ok in staging if this is 3.7.4 but in this mey be not true in 3.8-dev or 3.9-dev where language packs may not require changes as these branches are more focused on gaining stability and preparing for as easy as possible update to 4.0.

richard67 commented:
@wojsmol I don't understand. You mean with 3.8-dev and 3.9-dev language packs will provide updates for future versions? I do not see that, so I still think it is OK to exclude them.

wojsmol commented:
This cen be true, but for details I request comment from @mbabker or @infograf768

richard67 commented:
I have not tested this item.<br /><br />Changing back to not tested until it is clarified if language packs shall be excluded or not.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

wojsmol commented:
@richard67 I try contacting @infograf768  on Skype. Weiting on invitation aproval.

richard67 commented:
@wojsmol Have you contacted @infograf768 ? What did he say? Does the functionality introduced with this PR here have to check language packs, too? And for which use case?

@wilsonge What did you have in mind regarding language packs when this task was set up at JAB?

I ask because if language packs shall be checked, the recent change in this PR for filtering them out has to be reverted (without reverting the corrections of my own mistakes when I helped here a bit).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

wojsmol commented:
@richard67 No invitation aproval soo far.

wojsmol commented:
@richard67 JM is repleaing on github so give him more time.

wojsmol commented:
@piotrmocko @richard67 I discuss language packs matter with @infograf768 on skype. Summing up the discussion, special treatment for language packages will be required. Let's wait for his comments after reading this PR thoroughly. Currently he has internet problems and can not access joomla.org and github.com.

infograf768 commented:
Folks, I think indeed that core language packs (and possibly other specific extension language packs like weblinks) must keep their own update way. As you know, we have a cron job to prepare the update xmls automatically.

The reason is simple.

In any specific MAJOR J version (3.x, 4.x) some language packs may not be updated to last J version (because the Translation Team is late or has disapeared and is not replaced).

But, although such lang packs are not up to date, they are still used.
Example for 3.x : Albanian, Armenian, Dari, Hebrew, Uyghur, for the ones which are quite old.

Most of the time, users are specially concerned by the frontend and it is the frontend part of our packs which changes the less.
When they really need something absent or wrong used in the frontend, they create an override.

afaik, we will have a new repo for 4.x lang packs with totally new packs, but the situation may not change regarding that aspect.

richard67 commented:
@infograf768 Do I understand you right that this PR here is ok like it is, i.e. not showing core extensions and also not showing any language packs and their extensions (regardless if core or not)?

As you and @wojsmol hopefully understood, this PR here is for showing compatibility with a selected version, which will be the update offered for the core. E.g. if you are on 3.7.3 and now will be offered the update to 3.7.4, you will be able to see (when the complete functionality which is descibed in the description of this PR is ready) in the Joomla Update component which extensions are compatible to this 3.7.4 and which are not and which we don't know about.

So from my point of view it indeed does not make sense to handle language packs here - but @wojsmol did not want to believe me.

@infograf768 Please confirm if I did understand you right.

infograf768 commented:
don't know if pr is ok. would need precise test instructions while using staging, but in any case, users should be able to decide if they want to update or not depending on the infos they would get through this patch.

richard67 commented:
I have tested this item :white_check_mark: successfully on 2300d9770b92d1e186350f5d2a63b8725dcadf4e<br /><br />Alter test result again after previous discussion.
My successful test is sill valid.
Why jenkins and appveyor tests fail I don't know.
But for new PRs, jenkins is not used, so maybe it is related to that and not to this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16494">issues.joomla.org/tracker/joomla-cms/16494</a>.</sub>

wilsonge commented:
I leave language packs up for JM. Basically the idea here is to reproduce the interface https://github.com/nicksavov/joomla-cms/tree/2.5.x-with-pre-upgrade-check which was about the ability to see what's outdated and what's not for a major version upgrade (but we'd also be displaying it for all upgrades, major, minor patches). The issues there was that it based itself on getting the compatibility info from the extensions manifest which was a bad idea because it meant releasing extensions + templates for every CMS release. If you get it from the update manifest you don't force people to release a new extension just to update the manifest file compatibility.

As to whether language files are displayed in the list or not I'm unsure. Many language files aren't updated every patch release, sometimes every minor and sometimes less frequently than that. And having an out of date language pack is often not an update blocker (because as JM rightly points out the frontend is normally not so affected).

brianteeman commented:
I dont know why languages are even being discussed here. The PR is called **Allow to check 3rd party extensions compatibility before J! upgrade #16494**. Languages are not **3rd party**, they are never ALL 100% up to date and they wont break an upgrade

richard67 commented:
That was exactly my opinion. It was @wojsmol who forced that discussion.

wilsonge commented:
OK we're all good then :)

wilsonge commented:
Oops I accidenlty merged it to staging, so reverted it back and applied to 3.9-dev with https://github.com/joomla/joomla-cms/commit/5d919c3eaebfd010e68b383d350da8a130f86913

SamuelSchepp commented:
Hello, we at Technische Hochschule Mittelhessen, Gießen, Germany are currently developing a PR to support all features shown in the Mockup-screenshot based on 3.9-dev branch.

We will send in a PR early next week.

brianteeman commented:
@SamuelSchepp I am confused by your comment as this has already been merged into 3.9

wilsonge commented:
This was just the model's required. Their going to work on the full interface that was originally envisaged by Nick Savov + Andrew Eddie

brianteeman commented:
ah - thanks for the explanation
